### PR TITLE
Detect HTTP/1.1 on HTTP/2 connection

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/dotnet/aspnetcore/issues/14884

image

What do we think? It's not perfect, but it should be a lot more informative in most cases. And extra work only happens if request is already been rejected so has no perf impact. Will add unit tests if this idea has merit.

Also, we can do something similar for HTTP/1.1 (on invalid request line, check if the content is an HTTP/2 connection preface)

### CODE REVIEW

**Comment by Tratcher** [comment]
This doesn't seem much better than adding more content to Http2ErrorInvalidPreface without the additional checks.

A lot of the people that run into this aren't even looking at the server logs, they only notice a connection abort on the client. If we detected HTTP/1.x would it be possible to send back a hard coded 400 response before aborting the connection?

**Comment by TanayParikh** [comment]
Needs a full re-run given #39361.

**Comment by JamesNK** [comment]
Is sending back HTTP/1.1 on a HTTP/2 connection be legal? Sure, the client is wrong, but do two wrongs make a right?

Changing the server response is a much bigger piece of work. To do that I think this would need to be completely robust (i.e. when detecting the error, read additional content until either the end of the line or max request line length is reached). Then send a valid HTTP/1.1 response. That's a lot more work.

Agree that returning an error to the client would help more people, but this is a start.

**Comment by Tratcher** [comment]
> Is sending back HTTP/1.1 on a HTTP/2 connection be legal? Sure, the client is wrong, but do two wrongs make a right?

But it's not an HTTP/2 connection, is it? This check says the client is trying to do HTTP/1.x. The server hoped for HTTP/2 but didn't get it.

**Comment by JamesNK** [comment]
Ok.

I want to understand what you expect here as the ideal solution. It is more work than I expected to do so I want to make sure it is done right the first time and I don't waste time doing something that people don't want.

Proposal:
1. Read connection to prefix length
2. Check prefix
3. **New:** If invalid, instead of throwing connection exception, read connection until /n (or max request line length)
4. **New:** Check if line is HTTP/1.x (how comprehensive do you want to the check to be? Right now I'm just checking for the trailing "HTTP/1.1" on the request line. Do we care if the method or path is invalid?)
5. **New:** If the request is HTTP/1.x then return a 400 response and end the connection
6. If not HTTP1/x (no newline bytes, or exceeded request line length, or no HTTP/1.x version at end of line) then throw connection error like it does today.

**Comment by Tratcher** [comment]
Sounds right. 

The other case to consider in this code path is `5. Client (TLS) calling no TLS port`, not for the same PR but keep it in mind while working here. If we don't detect HTTP/1.x then we could fall back to sniff for TLS frames (you only need ~5bytes to check for the basic structure). If that's a match then we could respond with a fatal TLS Alert like unexpected_message or illegal_parameter to make the error more searchable than a generic TCP RST. protocol_version is tempting, but that implies TLS protocol.

https://techcommunity.microsoft.com/t5/iis-support-blog/ssl-tls-alert-protocol-and-the-alert-codes/ba-p/377132

**Comment by JamesNK** [comment]
@Tratcher Updated. Needs tests and cache the hardcoded response. Otherwise, is this what you had in mind?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 1]
```diff
@@ -437,22 +504,43 @@ private async Task<bool> TryReadPrefaceAsync()
...
 
-    private static bool ParsePreface(in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined)
+    private bool ParseHttp1x(ReadOnlySequence<byte> buffer, out HttpVersion httpVersion)
     {
-        consumed = buffer.Start;
-        examined = buffer.End;
+        httpVersion = HttpVersion.Unknown;
 
-        if (buffer.Length < ClientPreface.Length)
+        var reader = new SequenceReader<byte>(buffer);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:511]: We should probably trim this to at most `Limits.MaxRequestLineSize` before calling `TryReadTo` similar to what we do here:

https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L179-L204

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs [hunk 2]
```diff
@@ -394,6 +395,17 @@ public void Http3GoAwayStreamId(string connectionId, long goAwayStreamId)
...
     }
 
+    [LoggerMessage(54, LogLevel.Debug, @"Connection id ""{ConnectionId}"": Invalid content received on connection. Possible incorrect HTTP version detected. Expected {ExpectedHttpVersion} but received {DetectedHttpVersion}.", EventName = "PossibleInvalidHttpVersionDetected", SkipEnabledCheck = true)]
+    private static partial void PossibleInvalidHttpVersionDetected(ILogger logger, string connectionId, string expectedHttpVersion, string detectedHttpVersion);
+
+    public void PossibleInvalidHttpVersionDetected(string connectionId, HttpVersion expectedHttpVersion, HttpVersion detectedHttpVersion)
+    {
+        if (_generalLogger.IsEnabled(LogLevel.Debug))
+        {
+            PossibleInvalidHttpVersionDetected(_generalLogger, connectionId, HttpUtilities.VersionToString(expectedHttpVersion), HttpUtilities.VersionToString(detectedHttpVersion));
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs:405]: Should this be using the "Microsoft.AspNetCore.Server.Kestrel.BadRequests" category?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 3]
```diff
@@ -415,9 +437,54 @@ private async Task<bool> TryReadPrefaceAsync()
...
+                            if (ParseHttp1x(readableBuffer, out var detectedVersion))
+                            {
+                                if (detectedVersion == HttpVersion.Http10 || detectedVersion == HttpVersion.Http11)
+                                {
+                                    Log.PossibleInvalidHttpVersionDetected(ConnectionId, HttpVersion.Http2, detectedVersion);
+
+                                    // TODO: Cache bytes
+                                    await _context.Transport.Output.WriteAsync(Encoding.UTF8.GetBytes(@"HTTP/1.1 400 Bad Request
+Connection: close
+
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:474]:
```suggestion
Connection: close
Content-Type: text/plain
Content-Length: 57

An HTTP/1.x request was sent to an HTTP/2 only endpoint.
```

**Comment by Tratcher** [comment]
Exactly, it gives us a chance to tell the client what they did wrong without them having to go dig through the server logs.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 3 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:474]: What about server and date headers? Is it ok to leave them out? They're both dynamic: the date changes and server header can be disabled with configuration.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.cs]: Done

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Done

**Comment by JamesNK** [comment]
Why isn't a connection error thrown from matching the preface if the reader ends before reading up to the preface length? In that case `TryReadPrefaceAsync` returns false, and the connection sends a GOAWAY of NO_ERROR.

Is that against the spec - https://httpwg.org/specs/rfc7540.html#rfc.section.3.5 - Bug?

I updated `TryReadPrefaceAsync` to throw Http2ConnectionErrorException with a PROTOCOL_ERROR instead of returning false. It seems like this method doesn't need to return a boolean.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs]: Server is optional, let's exclude it for simplicity. Date is technically required, but with many exclusions for error cases, so I wouldn't worry about it right now.

https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.2

**Comment by Tratcher** [comment]
> Why isn't a connection error thrown from matching the preface if the reader ends before reading up to the preface length? In that case `TryReadPrefaceAsync` returns false, and the connection sends a GOAWAY of NO_ERROR.
> 
> Is that against the spec - https://httpwg.org/specs/rfc7540.html#rfc.section.3.5 - Bug?
> 
> I updated `TryReadPrefaceAsync` to throw Http2ConnectionErrorException with a PROTOCOL_ERROR when `ReadResult.IsComplete` is true.

If the reader ends then the connection is likely already closed. It's also common for port scanners to connect but not send any data. Neither case requires a response, or an exception log.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 4]
```diff
@@ -415,15 +449,60 @@ private async Task<bool> TryReadPrefaceAsync()
...
+                            {
+                                Log.PossibleInvalidHttpVersionDetected(ConnectionId, HttpVersion.Http2, detectedVersion);
+
+                                var responseBytes = InvalidHttp1xErrorResponseBytes ??= Encoding.ASCII.GetBytes(@"HTTP/1.1 400 Bad Request
+Connection: close
+Content-Type: text/plain
+Content-Length: 57
+
+An HTTP/1.x request was sent to an HTTP/2 only endpoint.
+");
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:481]: If this builds on linux are you going to get the right line endings? Make the \r\n explicit.

Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:481]: (checking my math)
```suggestion
Content-Length: 56

An HTTP/1.x request was sent to an HTTP/2 only endpoint.");
```

**Review by Tratcher — APPROVED** [review]

**Comment by JamesNK** [comment]
> > Why isn't a connection error thrown from matching the preface if the reader ends before reading up to the preface length? In that case `TryReadPrefaceAsync` returns false, and the connection sends a GOAWAY of NO_ERROR.
> > Is that against the spec - [httpwg.org/specs/rfc7540.html#rfc.section.3.5](https://httpwg.org/specs/rfc7540.html#rfc.section.3.5) - Bug?
> > I updated `TryReadPrefaceAsync` to throw Http2ConnectionErrorException with a PROTOCOL_ERROR when `ReadResult.IsComplete` is true.
> 
> If the reader ends then the connection is likely already closed. It's also common for port scanners to connect but not send any data. Neither case requires a response, or an exception log.

Ok. FYI today the server returns a GOAWAY frame with NO_ERROR. If you think nothing should be sent then that would be a change. I'm going to leave current behavior as-is and update my tests in this area.

**Comment by Tratcher** [comment]
Have you tried it with HttpClient to make sure it works as expected?

**Comment by JamesNK** [comment]
Yes, but not in an integration test. I'll see if I can add one.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs [hunk 5]
```diff
@@ -415,9 +449,55 @@ private async Task<bool> TryReadPrefaceAsync()
...
+                            else
+                            {
+                                state ^= ReadPrefaceState.Preface;
+                            }
+                        }
+                    }
+
+                    if (state.HasFlag(ReadPrefaceState.Http1x))
+                    {
+                        if (ParseHttp1x(readableBuffer, out var detectedVersion))
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:469]: Nit: It's not super obvious from the name that `ParseHttp1x` returns `true` when the `MaxRequestLineSize` is exceeded before an `\r` could be found. Maybe `ParseHttp1x` should just take the current `ReadPrefaceState` and return the updated one or something.

**Pull request merged**